### DISCUSSION TITLE
Question on Overlapping Extension and CDR matching

### DISCUSSION THREAD

**Discussion by firdaussoberi** [massimo-zaniboni/asterisell-v5#16] [Q&A] [answered]
Hi,

I just got a quick question regarding extension and matching of the billing in CDR. 
Say I have CompanyA and CompanyB, and say both companies requested the same extension number (it could be different employee but by request, the extension number is coincidentally same. Also, from customer's point of view, the extension is unique to them so it make perfect sense to start the running number again).

So I could have for example:
`CompanyA/20001`
`CompanyB/20001`

This is fine and I could add it through Entities->Customer section.
However, for CDR importing and billing, how would the importing engine knows which company is which?

Currently I use `asterisell-standard__v1` to import them, and say I import file `data.manual-importing__asterisell-standard__v1` , containing these lines (say listening from two different PBX but aggregately sent to import, and say with an overlapping extension):

> 2022-01-14 13:37:12,\N,outgoing,false,ANSWERED,283,20001,12067878458,,SIP-general-operator,
> 2022-01-14 15:17:19,\N,outgoing,false,ANSWERED,9,20001,12067878458,,SIP-general-operator

correponding to:
>
      callDate, toCallDate, direction, isRedirect, disposition, billsec, internalExtension,
      externalTelephoneNumber, portedTelephoneNumber, vendorAndChannelDomain, expectedCost 

as in `rating_tools/rate_engine/src/Asterisell/Cdr.hs`
with the first one representing `CompanyA/20001` and second one representing call from `CompanyB/20001`, then it would not know the difference since there is no companyID or organizationID stated right? Is there a format which could effectively distinguish scenario where same extension is used for different companies?

One solution could be to create another instance to separate the overlapping company/extension, but I would prefer to avoid this route as this is not ideal for each time overlapping extension is created. Could you guide us?

Thanks.

**Comment by massimo-zaniboni** [comment, marked as answer]
If you generate this datafile, you can use "virtual" extension like "companyA-12067878458", i.e. you generate CDR line like 

```
2022-01-14 13:37:12,\N,outgoing,false,ANSWERED,283,20001,companyA-12067878458,,SIP-general-operator,
```

Then when you insert the extension, you associate to the virtual ID "companyA-12067878458" the human readable name "12067878458", so your customer will see this human readable name. You can also say that all extensions like "companyA-*" will be associated to the customer.

Another way, it is using a different format containing info about the customer. But it requires the creation of new Haskell code probably.

**Reply by firdaussoberi** [reply]
Hi, thanks. Could you elaborate more on the virtual ID, when you say virtual extension `companyA-*`? Actually, 12067878458 is the outgoing number the extension will call for example, and `20001` is the extension that calls this number.

Would this then be `companyA-20001`? Sorry, but I'm a bit confused. 

Also, is there a reference I can look up on how to set the virtual extension?

Thank you so much.

**Reply by firdaussoberi** [reply]
I think I am starting to see what you said here. Do you mean instead of generating this line:

>..., 20001, 12067878458,....

with 20001 as the caller extension (that belongs to companyA) and 12067878458 as the outgoing receiver, I could instead maintain a reference table (somehow mapped like in DB table etc with associating say `20001` to `companyA-20001`) 

and then generate this line instead? 
>..., companyA-20001, 12067878458,....

This I could set in the name of the extension (the `Fix` function) as in here (in the unique extension codes) right? 
<img width="962" alt="Screenshot 2022-01-24 at 7 59 58 PM" src="https://user-images.githubusercontent.com/8426299/150779358-a7926261-dcde-4dad-b1d0-dc9441d9e630.png">

I think I'm starting to realize now that the extensions the CDR import engine refer to are actually the extension codes, these are stored in `ar_organization_unit_has_structure` table.

And regarding CDR, will this automatically identify the virtual extension to the specific customer, or do I have to meddle with the Haskel code?

Thanks again.

**Reply by firdaussoberi** [reply]
I just tried, that works! I can see the CDR entries of both companies now in `Status->Calls` , despite having same extension number. The trick is in the unique extension code. Thanks for the help! 😄
